libcudacxx/include/cuda/__bit/bitfield.h

```diff
@@ -0,0 +1,117 @@
+  return __dest;
+}
+
+_CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_DEVICE uint32_t __bfe(uint32_t __value, int __start, int __width) noexcept
+{
+  uint32_t __ret;
+  asm("bfe.u32 %0, %1, %2, %3;" : "=r"(__ret) : "r"(__value), "r"(__start), "r"(__width));
+  return __ret;
+}
+
+_CCCL_NODISCARD _CCCL_HIDE_FROM_ABI _CCCL_DEVICE uint64_t __bfe(uint64_t __value, int __start, int __width) noexcept
+{
+  uint64_t __ret;
+  asm("bfe.u64 %0, %1, %2, %3;" : "=l"(__ret) : "l"(__value), "r"(__start), "r"(__width));
+  return __ret;
+}
+
+#endif // defined(__CUDA_ARCH__)
+
+template <typename _Tp>
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp
+bitfield_insert(const _Tp __dest, const _Tp __source, int __start, int __width) noexcept
+{
+  static_assert(_CUDA_VSTD::__cccl_is_unsigned_integer_v<_Tp>, "bitfield_insert() requires unsigned integer types");
+  constexpr auto __digits = _CUDA_VSTD::numeric_limits<_Tp>::digits;
+  _CCCL_ASSERT(__width > 0 && __width <= __digits, "width out of range");
+  _CCCL_ASSERT(__start >= 0 && __start < __digits, "start position out of range");
+  _CCCL_ASSERT(__start + __width <= __digits, "start position + width out of range");
+  if constexpr (sizeof(_Tp) <= sizeof(uint64_t))
+  {
+    if (!_CUDA_VSTD::__cccl_default_is_constant_evaluated())
+    {
+      // clang-format off
+      NV_DISPATCH_TARGET( // all SM < 70
+        NV_PROVIDES_SM_70, (;),
+        NV_IS_DEVICE,      (using _Up = _CUDA_VSTD::_If<sizeof(_Tp) <= sizeof(uint32_t), uint32_t, uint64_t>;
+                            return ::cuda::__bfi(static_cast<_Up>(__dest), static_cast<_Up>(__source),
+                                                 __start, __width);))
+      // clang-format on
+    }
+  }
+  auto __mask = ::cuda::bitmask<_Tp>(__start, __width);
+  return ((__source << __start) & __mask) | (__dest & ~__mask);
+}
+
+template <typename _Tp>
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp
+bitfield_extract(const _Tp __value, int __start, int __width = 1) noexcept
+{
+  static_assert(_CUDA_VSTD::__cccl_is_unsigned_integer_v<_Tp>, "bitfield_extract() requires unsigned integer types");
+  constexpr auto __digits = _CUDA_VSTD::numeric_limits<_Tp>::digits;
+  _CCCL_ASSERT(__width > 0 && __width <= __digits, "width out of range");
+  _CCCL_ASSERT(__start >= 0 && __start < __digits, "start position out of range");
+  _CCCL_ASSERT(__start + __width <= __digits, "start position + width out of range");
+  if constexpr (sizeof(_Tp) <= sizeof(uint32_t))
+  {
+    if (!_CUDA_VSTD::__cccl_default_is_constant_evaluated())
+    {
+      // clang-format off
+      NV_DISPATCH_TARGET( // all SM < 70
+        NV_PROVIDES_SM_70, (;),
+        NV_IS_DEVICE,      (using _Up = _CUDA_VSTD::_If<sizeof(_Tp) <= sizeof(uint32_t), uint32_t, uint64_t>;
+                            return ::cuda::__bfe(static_cast<_Up>(__value), __start, __width);))
+      // clang-format on
+    }
+  }
+  return __value & ::cuda::bitmask<_Tp>(__start, __width);
```

Comment:
I don't think this resembles the implementation of `bfe` (see my comment related to the corresponding test):

> `bfe` is supposed to return the extracted bit field beginning from the LSB (see [reference](https://docs.nvidia.com/cuda/parallel-thread-execution/#integer-arithmetic-instructions-bfe))